prow/bugzilla/client.go

```diff
@@ -117,6 +119,105 @@ func (c *client) GetClones(bug *Bug) ([]*Bug, error) {
 	return getClones(c, bug)
 }
 
+// Gets children clones recursively using a mechanism similar to bfs
+func getRecursiveClones(c Client, root *Bug) ([]*Bug, error) {
+	var errs []error
+	var bug *Bug
+	clones := []*Bug{}
+	childrenQ := []*Bug{}
+	childrenQ = append(childrenQ, root)
+	// FYI Cannot think of any situation for circular clones
+	// But might need to revisit in case there are infinite loops at any point
+	for len(childrenQ) > 0 {
+		bug, childrenQ = childrenQ[0], childrenQ[1:]
+		clones = append(clones, bug)
+		children, err := getClones(c, bug)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("Error finding clones Bug#%d: %v", bug.ID, err))
+		}
+		if len(children) > 0 {
+			childrenQ = append(childrenQ, children...)
+		}
+	}
+	return clones, utilerrors.NewAggregate(errs)
+}
+
+// GetImmediateParents gets the Immediate parents of bugs with a matching summary
+func getImmediateParents(c Client, bug *Bug) ([]*Bug, error) {
+	var errs []error
+	parents := []*Bug{}
+	// One option would be to return as soon as the first parent is found
+	// ideally that should be enough, although there is a check in the getRoot function to verify this
+	// Logs would need to be monitored to verify this behavior
+	for _, parentID := range bug.DependsOn {
+		parent, err := c.GetBug(parentID)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("Failed to get parent bug #%d: %v", parentID, err))
+			continue
+		}
+		if parent.Summary == bug.Summary {
+			parents = append(parents, parent)
+		}
+	}
+	return parents, utilerrors.NewAggregate(errs)
+}
+
+func getRoot(c Client, bug *Bug) (*Bug, error) {
+	curr := bug
+	var errs []error
+	for len(curr.DependsOn) > 0 {
+		parent, err := getImmediateParents(c, curr)
+		if err != nil {
+			errs = append(errs, err)
+		}
+		if len(parent) > 0 {
+			// Might be a good idea to keep this check - since we dont expect more than 1 parent
+			if len(parent) > 1 {
+				errs = append(errs, fmt.Errorf("More than one parent found for bug #%d", curr.ID))
+			}
+			curr = parent[0]
+		} else {
+			break
+		}
+	}
+	return curr, utilerrors.NewAggregate(errs)
+}
+
+// GetRoot returns the original bug.
+func (c *client) GetRoot(bug *Bug) (*Bug, error) {
+	return getRoot(c, bug)
+}
+
+// GetAllClones returns all the clones of the bug including itself
+// Differs from GetClones as GetClones only gets the child clones which are one level lower
+func (c *client) GetAllClones(bug *Bug) ([]*Bug, error) {
```

Comment:
Can you put the explanation above in the `type Client interface`? The godoc here won't be visible from the outside as the `client` is unexported

Reply:
I might be missing something as I dont see any of the other functions in the interface having any explanations. Do you mean line 51 of this file or did you mean some other place?

Reply:
Yeah, I mean line 51. I don't know why it was done like that before, but having godocs on the implementation and not the interface doesn't really make any sense, as you can only see them when looking at the implementation. But the implementation is unexported, the only thing that is exported is the interface and that is returned when constructing a new client.

Reply:
Yup. Got it. Should I move the explanations to the interface for the rest of the functions as well?